src/persist/README.md

```diff
@@ -3,49 +3,194 @@
+  collection?
+- _persisted stream_: The above, but durably persisted to be replay-able across
+  restarts. Indexed access is available for the persisted records.
+- _(external) stream name_: A human-readable, process-unique `&str` identifier
+  for a persisted stream. Used to align persisted records across restarts. Also
+  used to allow read-only "tail"ing of an active persisted stream. Internally,
+  this stream name gets transparently mapped to a smaller internal stream id.
+- _blob_: A durable `&str` -> `[]byte` key value store. Persistence is
+  written in terms of a generic `Blob` and `Buffer` interface. Examples of blob
+  implementations include files in a folder and S3 (or other cloud storage).
+- _buffer_: A durable log with `[]byte` entries. Persistence is written in terms
+  of a generic `Blob` and `Buffer` interface. Examples of buffer implementations
+  include a traditional write-ahead log (possibly on EBS or similar) and Kafka.
+  TODO: Rename this to something better. Log?
+- _runtime_: The internally managed threads, doing the work of persisting new
+  records, compacting previously written ones, and managing read handles to
+  persistent data. Accessed externally via a thread-safe and clone-able
+  "client".
 
-### V1 and V2
+## External API
 
-`Persister` corresponds to [`PersisterV1` in the prototype][persister v1] and
-basically just supports synchronously writing data out, grabbing a snapshot of
-everything written so far, and unblocking compaction of everything less than a
-given timestamp. It could implemented on top of a SQLite table with something
-like the following schema:
+A persistence user initializes a `Buffer` and a `Blob` implementation and uses
+them to start the runtime, receiving a `RuntimeClient` in response. This client
+is the primary external interface to the persistence system. This runtime allows
+us to funnel everything a process is persisting through a single place for
+batching and rate limiting. The buffer and blob are exclusively locked for
+writes by this process (they can be opened for reads by other processes,
+however).
 
-[persister v1]: https://github.com/danhhz/differential-dataflow/blob/02673114b05933341893ab603327237a9583e432/persist/src/persister.rs#L30-L39
+The runtime stops when every copy of the client has been dropped (or when any
+client's stop method has been called), attempting to first gracefully finish any
+pending work, but rejecting new work.
 
-```sql
-CREATE TABLE persisted (persisted_id INT, key BYTES, val BYTES, ts BYTES, diff BYTES);
-CREATE INDEX ON persisted (persisted_id, ts, key);
+As needed, the user creates a persisted stream by name, returning a token. At
+dataflow construction time, this token can be exchanged to create one of several
+available persistence operators (see below). When the operator is run, it hands
+any blocking io or cpu heavy work to the runtime, asking the dataflow to be
```

Comment:
This reads like the persistence layer determines whether we use OCC for durability; can that eventually be made something that the coordinator determines?

Reply:
Is this the same usage of OCC as https://github.com/MaterializeInc/materialize/issues/7424#issuecomment-882886935? How do you see it fitting in here?

Reply:
Yes, sorry. Answered there, but same thing here: the compute plane could optimistically start doing the work, with the "is <time> persisted" bit blocking the eventual query return/sink emission. With this design, the persistence layer blocks that, which seems fine and safe to start and not at all a problem, but I could see eventually the coordinator or some other layer saying "I can handle that; it is my job".

Reply:
Ah gotcha! I've been using "pipelining" to talk about this but occ works too.

We can do that sort of thing if we need it, but my initial instinct is that everything is much simpler to reason about if anything handed to compute is known to be replayable, so that's where I planned to start. Sounds like you're on board with starting there, so for sure happy to revisit this later.

(If we did do this, how do you imagine that we'd model the "persistence frontier" vs the "compute frontier"? Does this apply if the persisted data is how we're separating storage and compute into different processes/machines?)

Reply:
Starting easy works great for me. The first Q was just about whether it was crucial for the design or not. 

Should we eventually do it, I think the coordinator already has this sort of information for timestamp bindings, and I think it is just a fair bit more logic that the coordinator feeds back to the boundaries (e.g. SELECT responses and SINK output). So, just two frontiers about 1. determined results, and 2. persisted results. Having one frontier seems much easier for now, and probably until there are complaints!

Reply:
Addressed this directly in the new version